### PR TITLE
Optimize NonlocalKernel and corresponding rigid body motion objects

### LINKED ISSUE #7837: Optimize jacobian calculation for NonlocalKernel & derived classes
### Description of the enhancement or error report
Provide additional optimization option in jacobian calculations in NonlocalKernels.
Possible optimization options:
  - Obtain jacobians from uo once per dof instead of at each qp
  - Execute nonlocal jacobian calculation only for dofs having nonzero contribution

### Rationale for the enhancement or information for reproducing the error
Reduce over all computational time. 

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
Nothing other than rigidbodymotion objects in phasefield.

### PR DESCRIPTION
**Pull request by SudiptaBiswas**
- Provides optimization option in NonlocalKernel
- Corrects advection velocity formulation
- Updates RBM kernels
- Corrects coupled variables in RBM kernel action
- Closes #7837

### CODE REVIEW

File: framework/src/kernels/NonlocalKernel.C [hunk 1]
```diff
@@ -39,6 +39,58 @@ NonlocalKernel::NonlocalKernel(const InputParameters & parameters) :
 }
 
 void
+NonlocalKernel::computeJacobian()
+{
```
Comment (permcody) [inline framework/src/kernels/NonlocalKernel.C]: Most of this code is copied directly out of Kernel. I wonder if there's a better way of implementing these algorithms in terms of the existing code. I realize that they are not 100% the same.

File: modules/phase_field/src/action/RigidBodyMultiKernelAction.C [hunk 2]
```diff
@@ -50,29 +50,44 @@ RigidBodyMultiKernelAction::act()
...
 
     //
+    // Create vector of order parameters
+    //
+
+    std::vector<VariableName> v;
+    unsigned int i = 0;
+    v.resize(_op_num);
+    for (unsigned int j = 0; j < _op_num; ++j)
+      v[i++] = _var_name_base + Moose::stringify(j);
```
Comment (permcody) [inline modules/phase_field/src/action/RigidBodyMultiKernelAction.C]: Why not use `j` on the lhs too?

File: modules/phase_field/src/action/RigidBodyMultiKernelAction.C [hunk 3]
```diff
@@ -50,29 +50,44 @@ RigidBodyMultiKernelAction::act()
...
+          arg[ind++] = _var_name_base + Moose::stringify(j);
     }
 
     //
+    // Create vector of order parameters
+    //
+
+    std::vector<VariableName> v;
+    unsigned int i = 0;
+    v.resize(_op_num);
```
Comment (permcody) [inline modules/phase_field/src/action/RigidBodyMultiKernelAction.C]: No need to have a resize operation here. You can size the vector as you create it.

**Review by permcody — COMMENTED** [review]
This looks mostly OK. I'll have @dschwen take a look too.

File: framework/src/kernels/NonlocalKernel.C (hunk 1 above)
Comment (SudiptaBiswas) [inline framework/src/kernels/NonlocalKernel.C]: I gave it some thought actually. My initial intention was to keep these in Kernel itself. However, for optimization purpose I had to alter few things and I did not want to disturb the Kernel class.

Comment (permcody) [inline framework/src/kernels/NonlocalKernel.C]: At the very least could you put documentation in each of these methods letting maintainers know that this code is copied from Kernel. That way if somebody comes in here and makes a change, they can check the other implementation as well.

Comment (SudiptaBiswas) [inline framework/src/kernels/NonlocalKernel.C]: Okay, will do that.

File: modules/phase_field/src/action/RigidBodyMultiKernelAction.C (hunk 2 above)
Comment (SudiptaBiswas) [inline modules/phase_field/src/action/RigidBodyMultiKernelAction.C]: Yeah, I should do that.

File: framework/include/kernels/NonlocalKernel.h [hunk 4]
```diff
@@ -34,8 +34,24 @@ class NonlocalKernel : public Kernel
 public:
   NonlocalKernel(const InputParameters & parameters);
 
+  /**
+   * computeJacobian and computeQpOffDiagJacobian methods are almost same
+   * as Kernel except for few additional optimization options regarding the integral terms.
+   * Looping order of _i & _j are reversed for providing opimization options and
+   * additional getUserObjectJacobian method is provided as an option to obtain
+   * jocobinas of the integral term from userobject once per dof
```
Comment (dschwen) [inline framework/include/kernels/NonlocalKernel.h]: "jocobinas"

File: framework/include/kernels/NonlocalKernel.h [hunk 5]
```diff
@@ -34,8 +34,24 @@ class NonlocalKernel : public Kernel
...
   virtual void computeJacobian();
   virtual void computeOffDiagJacobian(unsigned int jvar);
+
+  /**
+   * computeNonlocalJacobian and computeNonlocalOffDiagJacobian methods are
+   * introduced for providing the jacobian contribution corresponding to nolocal dofs.
+   * Nonlocal jacobian block is sized with respect to the all the dofs of jacobian variable.
+   * Looping order of _i & _k are reversed for opimization purpose and
+   * additional globalDoFEnabled method is provided as an option to execute nonlocal
+   * jocobina calculations only for nonlocal dofs that has nonzero jacobian contribution.
```
Comment (dschwen) [inline framework/include/kernels/NonlocalKernel.h]: "jocobina"

File: modules/phase_field/src/userobjects/ComputeGrainForceAndTorque.C [hunk 6]
```diff
@@ -78,7 +78,7 @@ ComputeGrainForceAndTorque::execute()
         for (_qp=0; _qp<_qrule->n_points(); ++_qp)
           if (_dF[_qp][j](0) != 0.0 || _dF[_qp][j](1) != 0.0 || _dF[_qp][j](2) != 0.0)
           {
-            const RealGradient compute_torque =_JxW[_qp] * _coord[_qp] * (_q_point[_qp] - centroid).cross(_dF[_qp][j]);
+            const RealGradient compute_torque =_JxW[_qp] * _coord[_qp] * (_current_elem->centroid() - centroid).cross(_dF[_qp][j]);
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/ComputeGrainForceAndTorque.C]: Why this change?

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/phase_field/src/userobjects/ComputeGrainForceAndTorque.C (hunk 6 above)
Comment (SudiptaBiswas) [inline modules/phase_field/src/userobjects/ComputeGrainForceAndTorque.C]: This helps in optimization while calculating the velocities later. Change in this userobject is done just to keep the formulation consistent everywhere.

**Review by dschwen — APPROVED** [review]

**Pull request merged**